chainerx_cc/chainerx/routines/linalg.cc

```diff
@@ -189,4 +192,161 @@ Array Inverse(const Array& a) {
+                Array mask = Eye(F.shape()[0], F.shape()[1], 0, Dtype::kBool, a.device());
+                F = Where(mask, INFINITY, F);
+                F = Reciprocal(F);
+
+                Array u_term{};
+                Array utgu = Dot(u.Transpose(), gu);
+                u_term = Dot(Dot(u, F * (utgu - utgu.Transpose())), sigma_mat);
+                if (m > k) {
+                    u_term = u_term + Dot(Dot(im - Dot(u, ut), gu), sigma_mat_inv);
+                }
+                u_term = Dot(u_term, vt);
+
+                Array v_term{};
+                Array vtgv = Dot(vt, gv);
+                v_term = Dot(Dot(sigma_mat, F * (vtgv - vtgv.Transpose())), vt);
+                if (n > k) {
+                    v_term = v_term + Dot(sigma_mat_inv, Dot(gvt, in - Dot(v, vt)));
+                }
+                v_term = Dot(u, v_term);
+
+                bctx.input_grad() = u_term + sigma_term + v_term;
+            });
+        }
+        bb.Finalize();
+    }
+
+    return std::make_tuple(std::move(u), std::move(s), std::move(vt));
+}
+
+Array PseudoInverse(const Array& a, float rcond) {
+    CheckRankTwoArray(a);
+    Dtype dtype = internal::GetMathResultDtype(a.dtype());
+    Array out = Empty(Shape({a.shape()[1], a.shape()[0]}), dtype, a.device());
+
+    {
+        NoBackpropModeScope scope{};
+
+        Array u{};
+        Array s{};
+        Array vt{};
+
+        std::tie(u, s, vt) = Svd(a, /*full_matrices=*/false, /*compute_uv=*/true);
+
+        Array cutoff = rcond * s.Max();
+        Array cutoff_indices = s <= cutoff;
+
+        Array sinv = Reciprocal(s);
+        sinv = Where(cutoff_indices, 0, sinv);
+
+        std::vector<ArrayIndex> indices{Slice{}, NewAxis{}};
+
+        a.device().backend().CallKernel<DotKernel>(vt.Transpose(), sinv.At(indices) * u.Transpose(), out);
+    }
+
+    // Reference:
+    // https://mathoverflow.net/questions/25778/analytical-formula-for-numerical-derivative-of-the-matrix-pseudo-inverse
+    {
+        BackwardBuilder bb{"pinv", a, out};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([a_tok = bb.RetainInput(0), out_tok = bb.RetainOutput(0), a_dtype = a.dtype(), &a_device = a.device()](
+                              BackwardContext& bctx) {
+                const Array& a = bctx.GetRetainedInput(a_tok);
+                const Array& out = bctx.GetRetainedOutput(out_tok);
+                const Array& gout = *bctx.output_grad();
+                Array I_a = Identity(a.shape()[0], a_dtype, a_device);
+                Array I_out = Identity(out.shape()[0], a_dtype, a_device);
+                bctx.input_grad() = (-Dot(Dot(out, gout.Transpose()), out) + Dot(Dot(Dot(out, out.Transpose()), gout), I_a - Dot(a, out)) +
+                                     Dot(Dot(Dot(I_out - Dot(out, a), gout), out.Transpose()), out))
+                                            .Transpose();
```

Comment:
Would splitting this into multiple lines (with temp variables) make it more readable?